# Base de Dados Breast Cancer
#### Análise se um tumor é maligno ou benigno através de diversos parâmetros.

### Pré-processamento dos Dados

In [1]:
# importando a biblioteca pandas do python
import pandas as pd

In [3]:
# colocando no 'dataframe' os atributos previsores
previsores = pd.read_csv('entradas_breast.csv', encoding = 'utf-8', sep = ',')

In [4]:
# colocando no 'dataframe' os atributos meta
classe = pd.read_csv('saidas_breast.csv', encoding = 'utf-8', sep = ',')

In [5]:
# visualizando os atributos 'previsores'
print(previsores)

      radius_mean   texture_mean  ...   symmetry_worst   fractal_dimension_worst
0           17.99          10.38  ...           0.4601                   0.11890
1           20.57          17.77  ...         275.0000                   0.08902
2           19.69          21.25  ...           0.3613                   0.08758
3           11.42          20.38  ...           0.6638                 173.00000
4           20.29          14.34  ...           0.2364                   0.07678
..            ...            ...  ...              ...                       ...
564         21.56          22.39  ...         206.0000                   0.07115
565         20.13          28.25  ...           0.2572                   0.06637
566         16.60          28.08  ...           0.2218                   0.07820
567         20.60          29.33  ...           0.4087                 124.00000
568          7.76          24.54  ...           0.2871                   0.07039

[569 rows x 30 columns]


In [6]:
# visualizando a dimensão dos atributos previsores
print(previsores.shape)

(569, 30)


Observe que existem 569 registros e 30 atributos previsores na base de dados.

In [7]:
# visualizando os atributos 'meta'
print(classe)

     0
0    0
1    0
2    0
3    0
4    0
..  ..
564  0
565  0
566  0
567  0
568  1

[569 rows x 1 columns]


In [8]:
# visualizando as dimensões do atributo 'meta'
print(classe.shape)

(569, 1)


Observe que para cada registro existe uma classe. Caso seja 0 indica que o paciente possui um tumor benigno, caso seja 1 indica que o paciente possui um tumor maligno.

In [9]:
# importando a biblioteca sklearn do python
from sklearn.model_selection import train_test_split

In [10]:
# separando a base de dados em treinamento e teste
prev_trein, prev_test, classe_trein, classe_test = train_test_split(previsores, classe, test_size = 0.25,
                                                                   random_state = 0)

In [11]:
# visualizando a dimensão de 'prev_trein'
print(prev_trein.shape)

(426, 30)


In [12]:
# visualizando a dimensão de 'prev_test'
print(prev_test.shape)

(143, 30)


In [13]:
# visualizando a dimensão de 'classe_trein'
print(classe_trein.shape)

(426, 1)


In [14]:
# visualizando a dimensão de 'classe_test'
print(classe_test.shape)

(143, 1)


### Configurando a Rede Neural Artificial

Vamos criar uma rede multicamada em que cada uma dessas camadas vão se ligar sequencialmente.

In [15]:
# importando a biblioteca keras do python
import keras

Using TensorFlow backend.


In [16]:
# importando recursos do keras
from keras.models import Sequential

In [17]:
# importando recursos do keras
from keras.layers import Dense

In [18]:
# criando um objeto 'clssificador'
classificador = Sequential()

In [19]:
# adicionando uma camada densa a rede neural
classificador.add(Dense(units = (((30 + 1) // 2) + 1), activation = 'relu', 
                        kernel_initializer = 'random_uniform', input_dim = 30))

In [21]:
# adicionando uma camada de saída na rede neural
classificador.add(Dense(units = 1, activation = 'sigmoid'))

In [23]:
# adicionando parâmetros para otimizar o cálculo de pesos
classificador.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['binary_accuracy'])

### Treinando a Rede Neural Artificial com os Dados Disponíveis

In [24]:
# treinando a rede neural
classificador.fit(prev_trein, classe_trein, batch_size = 10, epochs = 100)

Epoch 1/100
426/426 [==============================] - 0s 310us/step - loss: 8.6059 - binary_accuracy: 0.6056
Epoch 2/100
426/426 [==============================] - 0s 111us/step - loss: 2.1056 - binary_accuracy: 0.7183
Epoch 3/100
426/426 [==============================] - 0s 122us/step - loss: 0.7032 - binary_accuracy: 0.7958
Epoch 4/100
426/426 [==============================] - 0s 102us/step - loss: 0.5113 - binary_accuracy: 0.8427
Epoch 5/100
426/426 [==============================] - 0s 107us/step - loss: 0.4045 - binary_accuracy: 0.8545
Epoch 6/100
426/426 [==============================] - 0s 110us/step - loss: 0.4432 - binary_accuracy: 0.8357
Epoch 7/100
426/426 [==============================] - 0s 107us/step - loss: 0.3257 - binary_accuracy: 0.8897
Epoch 8/100
426/426 [==============================] - 0s 114us/step - loss: 0.3003 - binary_accuracy: 0.8826
Epoch 9/100
426/426 [==============================] - 0s 113us/step - loss: 0.2552 - binary_accuracy: 0.8944
Epoch 10/1

In [35]:
# importanda a biblioteca numpy do python
import numpy as np

In [38]:
# visualizando os pesos da camada de entrada
pesos0 = classificador.layers[0].get_weights()
# transformando a lista em um array numpy
pesos0 = np.asarray(pesos0)
# imprimindo os pesos de cada neurônio de entrada
print(pesos0[0].shape)
# imprimindo os pesos do neurônio bias implementado pelo algoritmo
print(pesos0[1].shape)

(30, 16)
(16,)


Observe que para cada um dos registros existem um peso diferente que se conectam com os 16 neurônios da camada densa, assim como o neurônio bias implementado.

In [39]:
# visualizando os pesos da primeira camada densa
pesos1 = classificador.layers[1].get_weights()
# transformando a lista em um array numpy
pesos1 = np.asarray(pesos1)
# imprimindo os pesos de cada neurônio da camada densa
print(pesos1[0].shape)
# imprimindo os pesos do neurônio bias implementado pelo algoritmo
print(pesos1[1].shape)

(16, 1)
(1,)


Observe que para cada um dos neurônios da camada densa a um peso que se conectam com um neurônio de saída, assim como o bias implementado na camada densa.

Observe que a rede neural conseguiu classificar 94.84% dos dados de treinamento.

A rede neural inicia com os pesos indicando um erro muito alto, e termina com um erro mais baixo.

### Avaliando a Capacidade de Previsão da Rede Neural Artificial

In [25]:
# realizando a predição com os dados de teste
previsoes = classificador.predict(prev_test)

In [26]:
# visualizando as predições obtidas pelo algoritmo
print(previsoes)

[[7.3834300e-02]
 [9.9339199e-01]
 [9.9581814e-01]
 [6.0816282e-01]
 [1.0000000e+00]
 [9.7623479e-01]
 [1.0000000e+00]
 [9.9551725e-01]
 [1.0000000e+00]
 [9.9999726e-01]
 [6.8063867e-01]
 [8.2249761e-01]
 [1.0000000e+00]
 [2.7091396e-01]
 [9.1690487e-01]
 [1.7346919e-02]
 [1.0000000e+00]
 [2.3794996e-07]
 [2.8357685e-02]
 [1.5947659e-14]
 [2.5851160e-02]
 [8.1124866e-01]
 [9.9573755e-01]
 [9.7955978e-01]
 [6.2270272e-01]
 [8.9941359e-01]
 [9.8625255e-01]
 [9.8872644e-01]
 [9.9599254e-01]
 [4.2041740e-10]
 [9.9766576e-01]
 [1.7536195e-11]
 [9.1778004e-01]
 [1.2663901e-03]
 [1.0000000e+00]
 [9.0932846e-04]
 [8.8793182e-01]
 [1.1771080e-05]
 [9.9429041e-01]
 [1.0337830e-03]
 [1.7322484e-01]
 [1.0000000e+00]
 [1.8643737e-03]
 [9.9656689e-01]
 [4.1336349e-01]
 [2.0287580e-16]
 [1.0000000e+00]
 [9.9238175e-01]
 [9.9984956e-01]
 [4.9100757e-02]
 [1.0000000e+00]
 [9.9945438e-01]
 [3.0710780e-07]
 [9.9952877e-01]
 [9.9216485e-01]
 [8.1826800e-01]
 [1.0000000e+00]
 [9.9986839e-01]
 [9.9775422e-0

Os resultados foram números próximos ou iguais a 0 e números próximos ou iguais a 1. Para as primeiras predições os resultados indicam que o câncer é benigno e para as segundas predições o algoritmo indicou o câncer como maligno.

In [27]:
# criando uma lista booleana para configurar as previsões como 0 ou 1
previsoes = (previsoes > 0.5)
# definindo um limiar de 0.5 para analisar os dados

Vamos analisar a acurácia do algoritmo e uma matriz de confusão.

In [28]:
# importando recursos do sklearn 
from sklearn.metrics import confusion_matrix, accuracy_score

In [29]:
# visualizando a precisão do algoritmo referente a base de dados de teste
precisao = accuracy_score(classe_test, previsoes)
print(precisao)

0.9230769230769231


O algoritmo obteve uma precisão de 92.30%.


In [30]:
# visualizando a matriz de confusão para os dados de teste
matriz = confusion_matrix(classe_test, previsoes)
print(matriz)

[[46  7]
 [ 4 86]]


Observe que 46 registros foram avaliados corretamente como câncer benigno, juntamente com 86 registros analisados como câncer maligno. Entretanto, 7 registros referentes a câncer benigno foram avaliados como maligno, assim como 4 registros de câncer maligno foram avaliados incorretamente como benignos.

In [31]:
# analisando a acurâcia e a perda usando o keras
resultado = classificador.evaluate(prev_test, classe_test)
print(resultado)

143/143 [==============================] - 0s 125us/step
[0.3743604321840343, 0.9230769276618958]


### Alguma Dúvida? Entre em Contato Comigo:

- [Me envie um e-mail](mailto:alysson.barbosa@ee.ufcg.edu.br);